sql/src/main/java/io/crate/analyze/symbol/FunctionCopyVisitor.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze.symbol;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Base class which can be used to create a visitor that has to replace functions.
+ *
+ * @param <C> context class for the visitor.
+ */
+public abstract class FunctionCopyVisitor<C> extends SymbolVisitor<C, Symbol> {
+
+    /**
+     * Traverses the functions arguments using {@link #process(Symbol, Object)}.
+     * If any process call returns a different instance then a new Function instance is returned.
+     */
+    protected Function processAndMaybeCopy(Function func, C context) {
+        List<Symbol> args = func.arguments();
+        switch (args.size()) {
+            // specialized functions to avoid allocations for common cases
+            case 1:
+                return oneArg(func, context);
+
+            case 2:
+                return twoArgs(func, context);
+
+            default:
+                return manyArgs(func, context);
```

Comment:
I've used this same pattern in the semi-join PR:

On a quick test:

```
private Function nestedFunction = expressions.asSymbol("1 + 1 + 1 + 1 + 1 * 1 + 1 + 1 + 1 + 1");

with optimization:
FuncSymbolsBenchmark.measureFuncMappingWithReplacement  avgt   50  140.898 ± 1.226  ns/op
FuncSymbolsBenchmark.measureFuncMappingWithout          avgt   50   89.718 ± 7.332  ns/op

without:
FuncSymbolsBenchmark.measureFuncMappingWithReplacement  avgt   50  405.221 ± 12.467  ns/op
FuncSymbolsBenchmark.measureFuncMappingWithout          avgt   50  332.942 ± 12.498  ns/op
```


I'll try to use this abstraction in other places (e.g. EvaluatingNormalizer) -
so that it actually will be worth the extra complexity. But good point about
pre-sizing the arrayList correctly.